### PR TITLE
Ley Lines timeline accordion

### PR DESCRIPTION
**Pull request by AkairyuGestalter**
Changes introduced:
1. Created a Ley Lines buff group for the Timeline to enable the user to see when they did not have Circle of Power during a given use of Ley Lines.
2. Added a Ley Lines usage accordion section, which displays the time of each use, the Circle of Power uptime percentage for each use, and expands to show the section of the timeline for each use.
3. Enforce display order of the Black Mage-specific sections
4. Update the patch support maximum for Black Mage
5. Adding my info as a contributor

### CODE REVIEW

**Comment by AkairyuGestalter** [comment]
.../analyse/PXATWBJ6ygknr978/15/1/ is a log of mine that you can use to see the leylines accordion in action. It basically does exactly the same work as the full fight timeline, just bound to the timeframe that the LeyLines buff is up.

File: src/parser/jobs/blm/Leylines.js [hunk 1]
```diff
@@ -68,26 +141,72 @@ export default class Leylines extends Module {
...
+			const thisCoPUptime = thisCoPHistory.map(cops => Math.max(cops.stop - cops.start, 0)).reduce((accumulator, currentValue) => accumulator + currentValue, 0)
+			// Note that since we're getting the actual duration, rather than the expected duration, technically we'll call it 100% uptime if you stay in the lines and die halfway through...
+			// However, since that'll get flagged as a morbid checklist item, that's probably ok.
+			const thisLeyLinesDuration = leyLinesEvent.stop - leyLinesEvent.start
+			const thisPercent = this._percentFunction(thisLeyLinesDuration, thisCoPUptime).toFixed(2)
+
+			return {
+				key: 'title-' + leyLinesEvent.start,
+				title: {
+					content: <Fragment>
```
Comment (ackwell) [inline src/parser/jobs/blm/Leylines.js]: Minor: can use `<>...</>` for fragments unless you need `key`s.

File: src/parser/jobs/blm/Leylines.js [hunk 2]
```diff
@@ -1,45 +1,117 @@
...
 			by: 'player',
 			abilityId: STATUSES.CIRCLE_OF_POWER.id,
 		}, this._onRemoveCircleOfPower)
+		this.addHook('removebuff', {
+			by: 'player',
+			abilityId: STATUSES.LEY_LINES.id,
+		}, this._onRemoveLeyLines)
 		this.addHook('complete', this._onComplete)
+
+		this._group = new Group({
```
Comment (ackwell) [inline src/parser/jobs/blm/Leylines.js]: I was questioning adding the new group at first, but looking at the example log, I can see how it'd be useful.

Perhaps down the line the group can be generalised or something, but it does the job nicely for now.

File: src/parser/jobs/blm/Leylines.js [hunk 3]
```diff
@@ -1,45 +1,117 @@
...
+			id: 'leybuffs',
+			content: 'Ley Lines Buffs',
+			order: 0,
+			nestedGroups: [],
+		})
+		this.timeline.addGroup(this._group)
+	}
+
+	// _onApplyBuff and _onLoseBuff retooled from RaidBuffs.js
+	_onApplyBuff(event) {
```
Comment (ackwell) [inline src/parser/jobs/blm/Leylines.js]: Minor; if i was being nitpicky, these shouldn't be `_on` functions, as they're not actually hooks. But weh. Naming.

File: src/parser/jobs/blm/Leylines.js [hunk 4]
```diff
@@ -68,26 +141,72 @@ export default class Leylines extends Module {
...
+				this._leyLines.current.stop = event.timestamps
+			}
+			this._leyLines.history.push(this._leyLines.current)
+		}
 		//check if there even were any events
 		if (!this._circleOfPowers.history.length) { return }
-		//filter out the last possible LL usage because it would make things weird.
-		this._circleOfPowers.history = this._circleOfPowers.history.filter(cops => cops.start < (event.timestamp - LEYLINE_DURATION))
-		this._leyLineHistory = this._leyLineHistory.filter(timestamps => timestamps < (event.timestamp - LEYLINE_DURATION))
+		if (!this._leyLines.history.length) { return }
```
Comment (ackwell) [inline src/parser/jobs/blm/Leylines.js]: Minor; could run these conditionals together

File: src/parser/jobs/blm/Leylines.js [hunk 5]
```diff
@@ -68,26 +141,72 @@ export default class Leylines extends Module {
...
+					percent: this._percentFunction(sumOfLeyLineDurations, sumOfCoPUpTime),
 				}),
 			],
 			//pretty random. Should be revised, maybe based on fights? 10% is ~ 1 GCD. So we allow that.
 			target: 90,
 		}))
 	}
+
+	output() {
+		const panels = this._leyLines.history.map(leyLinesEvent => {
```
Comment (ackwell) [inline src/parser/jobs/blm/Leylines.js]: Given the content is just the timeline link now, accordion might be overkill for the job. Perhaps a list of links or similar? Not sure.

**Pull request merged**